File: sdk/python/tests/test_budget_guard.py
"""Smoke test for the Python BudgetGuard wrapper.

Verifies that a Python BudgetGuard whose `check_before_llm` returns
``{"decision": "deny", ...}`` aborts ``session.send`` with a
RuntimeError that mentions "Budget exhausted" — the framework's
canonical denial signal — without the LLM ever being touched.

Run with:
    PYTHONPATH=python python tests/test_budget_guard.py
"""

from __future__ import annotations

import tempfile

from a3s_code import Agent, LocalWorkspaceBackend, PermissionPolicy, SessionOptions


INLINE_CONFIG = """
default_model = "anthropic/claude-sonnet-4-20250514"

providers "anthropic" {
  api_key = "test-key"
  models "claude-sonnet-4-20250514" {
    name = "Claude Sonnet 4"
  }
}
""".strip()


class DenyingGuard:
    """BudgetGuard that always denies the first LLM call and records
    everything the framework hands it for post-hoc assertions."""

    def __init__(self) -> None:
        self.llm_checks: list[tuple[str, int]] = []
        self.tool_checks: list[tuple[str, str]] = []
        self.llm_records: list[tuple[str, dict]] = []

    def check_before_llm(self, session_id: str, estimated_tokens: int) -> dict:
        self.llm_checks.append((session_id, estimated_tokens))
        return {
            "decision": "deny",
            "resource": "llm_tokens",
            "reason": "test cap exceeded",
        }

    def check_before_tool(self, session_id: str, tool_name: str) -> dict | None:
        self.tool_checks.append((session_id, tool_name))
        return None  # allow

    def record_after_llm(self, session_id: str, usage: dict) -> None:
        self.llm_records.append((session_id, usage))


class AllowingGuard:
    """BudgetGuard that returns None / no-op for every method. Verifies
    that "shape with no real methods" still works (the wrapper looks
    up by name at call time)."""

    def check_before_llm(self, session_id: str, estimated_tokens: int):
        return None

    def record_after_llm(self, session_id: str, usage: dict) -> None:
        pass


def main() -> None:
    workspace = tempfile.mkdtemp(prefix="a3s-budget-")
    agent = Agent.create(INLINE_CONFIG)

    # ----- Phase A: Deny -----
    guard = DenyingGuard()
    opts = SessionOptions()
    opts.permission_policy = PermissionPolicy(default_decision="allow")
    opts.workspace_backend = LocalWorkspaceBackend(workspace)
    opts.session_id = "budget-deny-test"
    opts.budget_guard = guard

    # Also exercise the getter — read-back must match what we wrote.
    assert opts.budget_guard is guard, "BudgetGuard getter must round-trip"

    session = agent.session(workspace, opts)

    try:
        _ = session.send("hello")
    except RuntimeError as exc:
        msg = str(exc)
        assert (
            "Budget exhausted" in msg or "llm_tokens" in msg
        ), f"expected budget-exhausted error, got: {exc!r}"
    else:
        raise AssertionError("send() must raise when BudgetGuard denies")

    assert len(guard.llm_checks) == 1, (
        f"check_before_llm must be consulted exactly once, got {guard.llm_checks!r}"
    )
    assert guard.llm_checks[0][0] == "budget-deny-test", (
        f"session_id must propagate, got {guard.llm_checks[0]!r}"
    )
    assert len(guard.llm_records) == 0, (
        f"record_after_llm must not fire when call was denied, got {guard.llm_records!r}"
    )

    # ----- Phase B: Allow / no-op shape -----
    # A guard with only allow-style methods must not break send().
    # We can't actually send without provider credentials, so we just
    # verify the SessionOptions roundtrip and that constructing a
    # session succeeds.
    allow_opts = SessionOptions()
    allow_opts.permission_policy = PermissionPolicy(default_decision="allow")
    allow_opts.workspace_backend = LocalWorkspaceBackend(workspace)
    allow_opts.session_id = "budget-allow-test"
    allow_opts.budget_guard = AllowingGuard()
    _ = agent.session(workspace, allow_opts)

    # ----- Phase C: clear back to None -----
    opts.budget_guard = None
    assert opts.budget_guard is None

    print("python sdk budget guard ok")


if __name__ == "__main__":
    main()

File: sdk/node/test_budget_guard.mjs
// Smoke test for the Node SDK BudgetGuard bridge.
//
// Verifies that a JS guard whose checkBeforeLlm returns
// { decision: 'deny', ... } aborts session.send before the LLM is
// touched. Runs with `node sdk/node/test_budget_guard.mjs`.

import assert from 'node:assert/strict'
import os from 'node:os'
import path from 'node:path'
import fs from 'node:fs'
import mod from './index.js'

const tmpRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'a3s-budget-'))
const workspace = path.join(tmpRoot, 'workspace')
fs.mkdirSync(workspace, { recursive: true })

const inlineConfig = `
default_model = "anthropic/claude-sonnet-4-20250514"

providers "anthropic" {
  api_key = "test-key"
  models "claude-sonnet-4-20250514" {
    name = "Claude Sonnet 4"
  }
}
`.trim()

const agent = await mod.Agent.create(inlineConfig)

const session = agent.session(workspace, {
  sessionId: 'budget-deny-node',
  permissionPolicy: { defaultDecision: 'allow' },
  workspaceBackend: new mod.LocalWorkspaceBackend(workspace),
})

let llmChecks = 0
let llmRecords = 0
let toolChecks = 0

session.setBudgetGuard({
  checkBeforeLlm: (ctx) => {
    llmChecks += 1
    assert.equal(ctx.sessionId, 'budget-deny-node', `wrong session_id, got ${ctx.sessionId}`)
    assert.equal(typeof ctx.estimatedTokens, 'number', 'estimatedTokens must be a number')
    return { decision: 'deny', resource: 'llm_tokens', reason: 'cap hit' }
  },
  recordAfterLlm: (_ctx) => {
    llmRecords += 1
  },
  checkBeforeTool: (_ctx) => {
    toolChecks += 1
    return null
  },
})

let threw = false
try {
  await session.send('hello')
} catch (err) {
  threw = true
  const msg = String(err).toLowerCase()
  assert.ok(
    msg.includes('budget exhausted') || msg.includes('llm_tokens'),
    `expected budget-exhausted error, got: ${err}`,
  )
}
assert.equal(threw, true, 'send() must throw when checkBeforeLlm denies')
assert.equal(llmChecks, 1, `checkBeforeLlm must fire exactly once, got ${llmChecks}`)
assert.equal(llmRecords, 0, `recordAfterLlm must not fire on Deny, got ${llmRecords}`)
assert.equal(toolChecks, 0, 'no tool was attempted; toolChecks must stay 0')

// Clearing the guard restores Allow-default behaviour. The mock LLM
// configured by `test_session_close.mjs` is not present here, so a
// real send would still fail at the provider level — we just verify
// that setBudgetGuard(null) is accepted without error.
session.setBudgetGuard(null)

// Throw, timeout, malformed-return, and host-process survival regressions run
// independently in test_callback_safety.mjs.

await session.closeAsync()
await agent.close()
fs.rmSync(tmpRoot, { recursive: true, force: true })

console.log('node sdk budget guard ok')
